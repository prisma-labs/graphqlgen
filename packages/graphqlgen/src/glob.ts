import * as glob from 'glob'
import { File } from 'graphqlgen-json-schema'
import { getPath } from './parse'

/**
 * Returns the path array from glob patterns
 */
export const extractGlobPattern = (file: File) => {
  return glob.sync(getPath(file))
}

/**
 * Handles the glob pattern of models.files
 */
export const handleGlobPattern = (files?: File[]): File[] => {
  if (!files) {
    return []
  }

  return files.reduce<File[]>((acc, file) => {
    const globedPaths = extractGlobPattern(file)

    if (globedPaths.length === 0) {
      return [...acc, file]
    }

    const globedFiles: File[] = globedPaths.map(path => {
      if (typeof file === 'string') {
        return path
      }

      return {
        path,
        defaultName: file.defaultName,
      }
    })

    return [...acc, ...globedFiles]
  }, [])
}
