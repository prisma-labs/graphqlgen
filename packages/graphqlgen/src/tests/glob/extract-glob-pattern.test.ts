import { extractGlobPattern } from '../../glob'

/**
 * Walk through all files under a directory
 */
test('Basic Walk', () => {
  const mockPaths = [
    './src/tests/fixtures/glob/dir-1/*.ts',
    './src/tests/fixtures/glob/dir-2/*.ts',
  ]
  expect(extractGlobPattern(mockPaths)).toMatchSnapshot()
})

/**
 * Walk through all folders under a dir using pattern
 */
test('Wild Walk', () => {
  const mockPaths = ['./src/tests/fixtures/glob/**/*.ts']
  expect(extractGlobPattern(mockPaths)).toMatchSnapshot()
})

/**
 * If no glob pattern is mentioned, Return the input array
 */
test('No Walk', () => {
  const mockPaths = [
    './src/tests/fixtures/glob/dir-1/file-11.ts',
    './src/tests/fixtures/glob/dir-2/file-21.ts',
  ]
  expect(extractGlobPattern(mockPaths)).toMatchSnapshot()
})

/**
 * Walk through mixed set of files
 */
test('Mixed Walk', () => {
  const mockPaths = [
    './src/tests/fixtures/glob/dir-1/file-11.ts',
    './src/tests/fixtures/glob/dir-2/file-21.ts',
    './src/tests/fixtures/glob/**/*.ts', // Wild Walk
    './src/tests/fixtures/glob/dir-1/*.ts', // Basic Walk
    './src/tests/fixtures/glob/dir-2/*.ts', // Basic Walk
  ]
  expect(extractGlobPattern(mockPaths)).toMatchSnapshot()
})

/**
 * Unknown Path, Returns an empty array
 */
test('Unknown Walk', () => {
  const mockPaths = ['./src/tests/fixtures/glob/dir-3/*.ts']
  expect(extractGlobPattern(mockPaths)).toMatchObject([])
})
