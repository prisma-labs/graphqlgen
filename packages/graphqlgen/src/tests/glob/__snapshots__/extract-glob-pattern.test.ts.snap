// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Walk 1`] = `
Array [
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-1/file-12.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
  "./src/tests/fixtures/glob/dir-2/file-22.ts",
]
`;

exports[`Mixed Walk 1`] = `
Array [
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-1/file-12.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
  "./src/tests/fixtures/glob/dir-2/file-22.ts",
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-1/file-12.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
  "./src/tests/fixtures/glob/dir-2/file-22.ts",
]
`;

exports[`No Walk 1`] = `
Array [
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
]
`;

exports[`Wild Walk 1`] = `
Array [
  "./src/tests/fixtures/glob/dir-1/file-11.ts",
  "./src/tests/fixtures/glob/dir-1/file-12.ts",
  "./src/tests/fixtures/glob/dir-2/file-21.ts",
  "./src/tests/fixtures/glob/dir-2/file-22.ts",
]
`;
